### PR TITLE
v0.30: Reimplement xkbswitch driver for keyboardlayout block issue #1512

### PR DESCRIPTION
**Pull request by tusharxoxoxo**
On raising this pull request, I confirm the following (please check boxes):

- [x] I have read and understood the [contributors guide](https://github.com/CCExtractor/ccextractor/blob/master/.github/CONTRIBUTING.md).
- [x] I have checked that another pull request for this purpose does not exist.
- [ ] I have considered and confirmed that this submission will be valuable to others.
- [x] I accept that this submission may not be used, and the pull request is closed at the will of the maintainer.
- [x] I give this submission freely and claim no ownership of its content.
- [ ] I have mentioned this change in the [changelog](https://github.com/CCExtractor/ccextractor/blob/master/docs/CHANGES.TXT).
My familiarity with the project is as follows (check one):

- [ ] I have never used CCExtractor.
- [x] I have used CCExtractor just a couple of times.
- [ ] I absolutely love CCExtractor, but have not contributed previously.
- [x] I am an active contributor to CCExtractor.

This PR adds an xkbswitch driver for the keyboard layout block.

There are a few things that aren't entirely done yet, though:
1. few changes needed to be made on line 117, I didn't understand exactly what it was doing
2. not sure whether I have implemented it rightly, someone please tell me how to do it properly
3. I was following what was done with setxkbmap
 
Any feedback/help is appreciated :-)

### CODE REVIEW

File: src/blocks/keyboard_layout.rs [hunk 1]
```diff
@@ -153,6 +155,33 @@ struct Info {
...
+        let mut parts = output.split(':');
+        let layout = parts.next().unwrap_or_default().trim().to_string();
+        let variant = parts.next().map(|v| v.trim().to_string());
+
+        Ok(Info { layout, variant })
+    }
+
+    async fn wait_for_change(&mut self) -> Result<()> {
+        // This backend doesn't support waiting for changes
+        Ok(())
```
Comment (MaxVerevkin) [inline src/blocks/keyboard_layout.rs:181]: Then it should `sleep()` (see `SetXkbMap`)

Comment (tusharxoxoxo) [reply src/blocks/keyboard_layout.rs:181]: thanks

Comment (tusharxoxoxo) [reply src/blocks/keyboard_layout.rs:181]: I have made the changes that you told me about. Are there any other changes I should make to make this working

**Comment by tusharxoxoxo** [comment]
why is pre-commit / pre-commit (clippy) (pull_request) Failing after 2m failed? How to fix this??

**Comment by MaxVerevkin** [comment]
> why is pre-commit / pre-commit (clippy) (pull_request) Failing after 2m failed? How to fix this??

Because the code doesn't compile:

```
error[E0277]: `?` couldn't convert the error to `errors::Error`
   --> src/blocks/keyboard_layout.rs:169:72
    |
169 |         let output = Command::new("xkbswitch").arg("-p").output().await?;
    |                                                                        ^ the trait `From<std::io::Error>` is not implemented for `errors::Error`
    |
    = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
    = help: the following other types implement trait `FromResidual<R>`:
              <std::result::Result<T, F> as FromResidual<Yeet<E>>>
              <std::result::Result<T, F> as FromResidual<std::result::Result<Infallible, E>>>
    = note: required for `std::result::Result<keyboard_layout::Info, errors::Error>` to implement `FromResidual<std::result::Result<Infallible, std::io::Error>>`

error[E0609]: no field `0` on type `&mut XkbSwitch`
   --> src/blocks/keyboard_layout.rs:180:20
    |
180 |         sleep(self.0 .0).await;
    |                    ^ unknown field

Some errors have detailed explanations: E0277, E0609.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `i3status-rs` due to 2 previous errors
```

To fix the first error, you need to have a `.error()` before `?` to add context to the error and convert it to the correct error type. Again, see `setxkbmap`.

Regarding the second error: you need to actually store the update interval in your struct.

File: src/blocks/keyboard_layout.rs [hunk 2]
```diff
@@ -113,6 +114,7 @@ pub async fn run(config: Config, mut api: CommonApi) -> Result<()> {
         KeyboardLayoutDriver::LocaleBus => Box::new(LocaleBus::new().await?),
         KeyboardLayoutDriver::KbddBus => Box::new(KbddBus::new().await?),
         KeyboardLayoutDriver::Sway => Box::new(Sway::new(config.sway_kb_identifier).await?),
+        KeyboardLayoutDriver::XkbSwitch => Box::new(XkbSwitch::new()),
```
Comment (MaxVerevkin) [inline src/blocks/keyboard_layout.rs:117]:
```suggestion
        KeyboardLayoutDriver::XkbSwitch => Box::new(XkbSwitch(config.interval)),
```

File: src/blocks/keyboard_layout.rs [hunk 3]
```diff
@@ -153,6 +155,39 @@ struct Info {
...
+#[async_trait]
+impl Backend for XkbSwitch {
+    async fn get_info(&mut self) -> Result<Info> {
+        let output = Command::new("xkbswitch")
+            .arg("-query")
+            .output()
+            .await
+            .error("Failed to execute xkbswitch command")?;
+            
+        let output = String::from_utf8_lossy(&output.stdout).error("xkbswitch produced a non-UTF8 output")?;
```
Comment (MaxVerevkin) [inline src/blocks/keyboard_layout.rs:169]: `from_utf8_lossy` is non-failable. It does not return a `Result`.

**Comment by ammgws** [comment]
What's this CCExtractor stuff in the first post? Copy-paste error?

**Pull request closed without merging**